Flaky Test: test list with prior objects

Has failed for both Blob and S3. 

```BlobStorageTest > test list with prior objects() FAILED
    kotlinx.coroutines.test.UncaughtExceptionsBeforeTest at TestScope.kt:238

kotlinx.coroutines.test.UncaughtExceptionsBeforeTest: There were uncaught exceptions before the test started. Please avoid this, as such exceptions are also reported in a platform-dependent manner so that they are not lost.
	at app//kotlinx.coroutines.test.TestScopeImpl.enter(TestScope.kt:238)
	at app//kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt.runTest-8Mi8wO0(TestBuilders.kt:308)
	at app//kotlinx.coroutines.test.TestBuildersKt.runTest-8Mi8wO0(Unknown Source)
	at app//kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt.runTest-8Mi8wO0(TestBuilders.kt:167)
	at app//kotlinx.coroutines.test.TestBuildersKt.runTest-8Mi8wO0(Unknown Source)
	at app//kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt.runTest-8Mi8wO0$default(TestBuilders.kt:159)
	at app//kotlinx.coroutines.test.TestBuildersKt.runTest-8Mi8wO0$default(Unknown Source)
	at app//xtdb.api.storage.ObjectStoreTest.test list with prior objects(ObjectStoreTest.kt:105)
	at java.base@21.0.8/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base@21.0.8/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base@21.0.8/java.util.ArrayList.forEach(ArrayList.java:1596)
	Suppressed: kotlinx.coroutines.channels.ClosedReceiveChannelException: Channel was closed
		at kotlinx.coroutines.channels.BufferedChannel.getReceiveException(BufferedChannel.kt:1762)
		at kotlinx.coroutines.channels.BufferedChannel.resumeWaiterOnClosedChannel(BufferedChannel.kt:2204)
		at kotlinx.coroutines.channels.BufferedChannel.resumeReceiverOnClosedChannel(BufferedChannel.kt:2193)
		at kotlinx.coroutines.channels.BufferedChannel.cancelSuspendedReceiveRequests(BufferedChannel.kt:2186)
		at kotlinx.coroutines.channels.BufferedChannel.completeClose(BufferedChannel.kt:1963)
		at kotlinx.coroutines.channels.BufferedChannel.isClosed(BufferedChannel.kt:2242)
		at kotlinx.coroutines.channels.BufferedChannel.isClosedForSend0(BufferedChannel.kt:2217)
		at kotlinx.coroutines.channels.BufferedChannel.isClosedForSend(BufferedChannel.kt:2214)
		at kotlinx.coroutines.channels.BufferedChannel.completeCloseOrCancel(BufferedChannel.kt:1935)
		at kotlinx.coroutines.channels.BufferedChannel.closeOrCancelImpl(BufferedChannel.kt:1828)
		at kotlinx.coroutines.channels.BufferedChannel.close(BufferedChannel.kt:1787)
		at kotlinx.coroutines.channels.SendChannel$DefaultImpls.close$default(Channel.kt:263)
		at xtdb.api.log.LocalLog$subscribe$subscription$1$1$3.invokeSuspend(LocalLog.kt:240)
		at xtdb.api.log.LocalLog$subscribe$subscription$1$1$3.invoke(LocalLog.kt)
		at xtdb.api.log.LocalLog$subscribe$subscription$1$1$3.invoke(LocalLog.kt)
		at kotlinx.coroutines.flow.FlowKt__EmittersKt.invokeSafely$FlowKt__EmittersKt(Emitters.kt:212)
		at kotlinx.coroutines.flow.FlowKt__EmittersKt.access$invokeSafely$FlowKt__EmittersKt(Emitters.kt:1)
		at kotlinx.coroutines.flow.FlowKt__EmittersKt$onCompletion$$inlined$unsafeFlow$1.collect(SafeCollector.common.kt:117)
		at kotlinx.coroutines.flow.FlowKt__EmittersKt$onCompletion$$inlined$unsafeFlow$1$1.invokeSuspend(Unknown Source)
		at _COROUTINE._BOUNDARY._(CoroutineDebugging.kt:42)
		at xtdb.api.log.LocalLog$subscribe$subscription$1$msg$1.invokeSuspend(LocalLog.kt:245)
		at kotlinx.coroutines.TimeoutKt.withTimeoutOrNull(Timeout.kt:102)
		at xtdb.api.log.LocalLog$subscribe$subscription$1.invokeSuspend(LocalLog.kt:245)
		Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [CoroutineId(135660), "coroutine#135660":StandaloneCoroutine{Cancelled}@6423cff6, Dispatchers.Default]
	Caused by: kotlinx.coroutines.channels.ClosedReceiveChannelException: Channel was closed
		at kotlinx.coroutines.channels.BufferedChannel.getReceiveException(BufferedChannel.kt:1762)
		at kotlinx.coroutines.channels.BufferedChannel.resumeWaiterOnClosedChannel(BufferedChannel.kt:2204)
		at kotlinx.coroutines.channels.BufferedChannel.resumeReceiverOnClosedChannel(BufferedChannel.kt:2193)
		at kotlinx.coroutines.channels.BufferedChannel.cancelSuspendedReceiveRequests(BufferedChannel.kt:2186)
		at kotlinx.coroutines.channels.BufferedChannel.completeClose(BufferedChannel.kt:1963)
		at kotlinx.coroutines.channels.BufferedChannel.isClosed(BufferedChannel.kt:2242)
		at kotlinx.coroutines.channels.BufferedChannel.isClosedForSend0(BufferedChannel.kt:2217)
		at kotlinx.coroutines.channels.BufferedChannel.isClosedForSend(BufferedChannel.kt:2214)
		at kotlinx.coroutines.channels.BufferedChannel.completeCloseOrCancel(BufferedChannel.kt:1935)
		at kotlinx.coroutines.channels.BufferedChannel.closeOrCancelImpl(BufferedChannel.kt:1828)
		at kotlinx.coroutines.channels.BufferedChannel.close(BufferedChannel.kt:1787)
		at kotlinx.coroutines.channels.SendChannel$DefaultImpls.close$default(Channel.kt:263)
		at xtdb.api.log.LocalLog$subscribe$subscription$1$1$3.invokeSuspend(LocalLog.kt:240)
		at xtdb.api.log.LocalLog$subscribe$subscription$1$1$3.invoke(LocalLog.kt)
		at xtdb.api.log.LocalLog$subscribe$subscription$1$1$3.invoke(LocalLog.kt)
		at kotlinx.coroutines.flow.FlowKt__EmittersKt.invokeSafely$FlowKt__EmittersKt(Emitters.kt:212)
		at kotlinx.coroutines.flow.FlowKt__EmittersKt.access$invokeSafely$FlowKt__EmittersKt(Emitters.kt:1)
		at kotlinx.coroutines.flow.FlowKt__EmittersKt$onCompletion$$inlined$unsafeFlow$1.collect(SafeCollector.common.kt:117)
		at kotlinx.coroutines.flow.FlowKt__EmittersKt$onCompletion$$inlined$unsafeFlow$1$1.invokeSuspend(Unknown Source)
		at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
		at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:98)
		at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:586)
		at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:829)
		at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:717)
		at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:704)```

--- a/core/src/main/kotlin/xtdb/api/log/LocalLog.kt
+++ b/core/src/main/kotlin/xtdb/api/log/LocalLog.kt
@@ -3,6 +3,7 @@
 package xtdb.api.log
 
 import kotlinx.coroutines.*
+import kotlinx.coroutines.cancel
 import kotlinx.coroutines.channels.Channel
 import kotlinx.coroutines.flow.*
 import kotlinx.coroutines.future.future
@@ -20,6 +21,7 @@ import xtdb.database.proto.localLog
 import xtdb.time.InstantUtil.asMicros
 import xtdb.time.InstantUtil.fromMicros
 import java.io.DataInputStream
+import java.io.IOException
 import java.nio.ByteBuffer
 import java.nio.channels.Channels
 import java.nio.channels.ClosedByInterruptException
@@ -238,11 +240,22 @@ class LocalLog(
                         }
                     }
                     .onCompletion { ch.close() }
+                    .catch {
+                        try {
+                            throw it
+                        } catch (_: ClosedByInterruptException) {
+                            throw CancellationException()
+                        } catch (_: InterruptedException) {
+                            throw CancellationException()
+                        }
+                    }
                     .collect()
             }
 
             while (true) {
-                val msg = withTimeoutOrNull(1.minutes) { ch.receive() }
+                val msg = withTimeoutOrNull(1.minutes) {
+                    ch.receiveCatching().let { if (it.isClosed) throw CancellationException() else it.getOrThrow()}
+                }
                 runInterruptible { subscriber.processRecords(listOfNotNull(msg)) }
             }
         }

--- a/core/src/test/kotlin/xtdb/api/log/LocalLogTest.kt
+++ b/core/src/test/kotlin/xtdb/api/log/LocalLogTest.kt
@@ -1,0 +1,41 @@
+package xtdb.api.log
+
+import kotlinx.coroutines.test.runTest
+import org.junit.jupiter.api.RepeatedTest
+import org.junit.jupiter.api.Tag
+import org.junit.jupiter.api.io.TempDir
+import xtdb.api.log.Log.Message
+import xtdb.api.log.Log.Record
+import java.nio.file.Path
+import kotlin.time.Duration.Companion.seconds
+
+class LocalLogTest {
+
+    @TempDir
+    lateinit var tempDir: Path
+
+    @Tag("integration")
+    @RepeatedTest(5000)
+    fun `close should cancel all subscription coroutines without leaking`() = runTest(timeout = 10.seconds) {
+        val log = LocalLog.Factory(tempDir.resolve("log")).openLog(emptyMap())
+
+        // Create a subscription
+        val records = mutableListOf<Record>()
+        val subscription = log.subscribe(
+            object : Log.Subscriber {
+                override val latestProcessedMsgId: Long = -1
+                override val latestSubmittedMsgId: Long = -1
+                override fun processRecords(recs: List<Record>) {
+                    recs.forEach { records.add(it) }
+                }
+            },
+            -1
+        )
+
+        log.appendMessage(Message.FlushBlock(1))
+
+        subscription.close()
+
+        log.close()
+    }
+}

